wrapper/src/main/java/software/amazon/jdbc/plugin/federatedauth/FederatedAuthPlugin.java

```diff
@@ -0,0 +1,317 @@
+  }
+
+
+  @Override
+  public Connection connect(
+      final String driverProtocol,
+      final HostSpec hostSpec,
+      final Properties props,
+      final boolean isInitialConnection,
+      final JdbcCallable<Connection, SQLException> connectFunc)
+      throws SQLException {
+    return connectInternal(hostSpec, props, connectFunc);
+  }
+
+  @Override
+  public Connection forceConnect(
+      final @NonNull String driverProtocol,
+      final @NonNull HostSpec hostSpec,
+      final @NonNull Properties props,
+      final boolean isInitialConnection,
+      final @NonNull JdbcCallable<Connection, SQLException> forceConnectFunc)
+      throws SQLException {
+    return connectInternal(hostSpec, props, forceConnectFunc);
+  }
+
+  private Connection connectInternal(HostSpec hostSpec, Properties props,
+      JdbcCallable<Connection, SQLException> connectFunc) throws SQLException {
+
+    checkIdpCredentialsWithFallback(props);
+
+    String host = IamAuthUtils.getIamHost(IAM_HOST.getString(props), hostSpec);
+
+    int port = IamAuthUtils.getIamPort(
+        StringUtils.isNullOrEmpty(IAM_DEFAULT_PORT.getString(props)) ? -1 : IAM_DEFAULT_PORT.getInteger(props),
+        hostSpec,
+        this.pluginService.getDialect().getDefaultPort());
+
+    final Region region = getRegion(host, props);
+
+    final String cacheKey = getCacheKey(
+        DB_USER.getString(props),
+        host,
+        port,
+        region);
+
+    final TokenInfo tokenInfo = tokenCache.get(cacheKey);
+
+    final boolean isCachedToken = tokenInfo != null && !tokenInfo.isExpired();
+
+    if (isCachedToken) {
+      LOGGER.finest(
+          () -> Messages.get(
+              "FederatedAuthPlugin.useCachedIamToken",
+              new Object[] {tokenInfo.getToken()}));
+      PropertyDefinition.PASSWORD.set(props, tokenInfo.getToken());
+    } else {
+      updateAuthenticationToken(hostSpec, props, region, cacheKey);
+    }
+
+    PropertyDefinition.USER.set(props, DB_USER.getString(props));
+
+    try {
+      return connectFunc.call();
+    } catch (final SQLException exception) {
+      updateAuthenticationToken(hostSpec, props, region, cacheKey);
+      return connectFunc.call();
+    } catch (final Exception exception) {
+      LOGGER.warning(
+          () -> Messages.get(
+              "FederatedAuthPlugin.unhandledException",
+              new Object[] {exception}));
+      throw new SQLException(exception);
+    }
+  }
+
+  private void checkIdpCredentialsWithFallback(final Properties props) {
+    if (IDP_USERNAME.getString(props) == null) {
+      IDP_USERNAME.set(props, PropertyDefinition.USER.getString(props));
+    }
+
+    if (IDP_PASSWORD.getString(props) == null) {
+      IDP_PASSWORD.set(props, PropertyDefinition.PASSWORD.getString(props));
+    }
+  }
+
+  private void updateAuthenticationToken(HostSpec hostSpec, Properties props, Region region, String cacheKey)
```

Comment:
Can we move these IAM specific methods to IamAuthUtils to reduce duplicated code?